[WIP]: Index payments by sequence number

carlaKC opened the pull request:
First draft of migration for indexing payments, see #4164 for details. 
```
payments-index-bucket
      <sequence nr>: <payment hash>
      <sequence nr>: <payment hash>
```

Since we have historic duplicate payments to the same hash, multiple entries in this index may **point to the same payment**. We take this into account when retrieving payments for a given hash by making sure that our sequence number matches what we're looking for. 

Last commit needs testing + current tests are failing, currently there as an example of how we would use the index.

joostjager commented on channeldb/payment_control_test.go:
```diff
@@ -350,7 +350,7 @@ func TestPaymentControlFailsWithoutInFlight(t *testing.T) {
 }
 
-// TestPaymentControlDeleteNonInFlight checks that calling DeletaPayments only
+// TestPaymentControlDeleteNonInFlight checks that calling DeletePayments only
```
I would squash the test with the changes

joostjager commented on channeldb/payments_test.go:
```diff
@@ -422,3 +423,63 @@ func TestQueryPayments(t *testing.T) {
 }
+
+// appendDuplicatePayment adds a duplicate payment to an existing payment. Note
+// that this function requires a unique sequence number.
```
The only reason we need this code is to reproduce a legacy database right? Maybe explain that in a comment

joostjager commented on channeldb/migration14/migration_test.go:
```diff
@@ -0,0 +1,125 @@
+		t.Run(test.name, func(t *testing.T) {
+			// Before the migration we have a payments bucket.
+			before := func(tx kvdb.RwTx) error {
+				return migtest.RestoreDB(
```
Nice to see another supporter of this style of migration test

joostjager commented on channeldb/db.go:
```diff
@@ -136,6 +137,11 @@ var (
+			// Migrate to index payments by sequence number.
+			number:    14,
+			migration: migration14.MigrateSequenceIndex,
+		},
```
Squash with previous?

joostjager commented on channeldb/invoices.go:
```diff
@@ -708,85 +708,46 @@ func (d *DB) QueryInvoices(q InvoiceQuery) (InvoiceSlice, error) {
-			}
+		// Create a paginater which reads from our add index bucket with
+		// the parameters provided by the invoice query.
+		paginater := newPagianter(
```
"Pagianter"

joostjager commented on channeldb/invoices.go:
```diff
@@ -708,85 +708,46 @@ func (d *DB) QueryInvoices(q InvoiceQuery) (InvoiceSlice, error) {
+		// maybeAddInvoice looks up an invoice based on the index we
+		// are given and returns the number of items we have added to
+		// our set of invoices.
+		maybeAddInvoice := func(_, indexValue []byte) (int, error) {
```
There must be a better name for this func

joostjager commented on channeldb/invoices.go:
```diff
@@ -708,85 +708,46 @@ func (d *DB) QueryInvoices(q InvoiceQuery) (InvoiceSlice, error) {
 			// At this point, we've exhausted the offset, so we'll
 			// begin collecting invoices found within the range.
 			resp.Invoices = append(resp.Invoices, invoice)
+			return 1, nil
```
If it only returns 0 and 1, it could be a bool return value

carlaKC replied:
I wanted to keep it generic in case we have a case where multiple items are appended. But perhaps that's not unlikely.

joostjager commented on channeldb/payments.go:
```diff
@@ -92,6 +92,14 @@ var (
+	// payment hash.
+	// payments-sequence-index-bucket
+	// 	|--sequence-number: <payment hash>
+	//      |--sequence-number: <payment hash>
```
two identical keys?

carlaKC replied:
Ah, wasn't clear on the schema for these comments. Wrapped in `<>` to indicate that they're keyed by value, not a string.

joostjager commented on channeldb/payment_control.go:
```diff
@@ -153,9 +153,22 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+		seqBytes := bucket.Get(paymentSequenceKey)
+		if seqBytes != nil {
+			indexBucket := tx.ReadWriteBucket(paymentsIndexBucket)
+			if err := indexBucket.Delete(seqBytes); err != nil {
```
Squash with previous commit to keep it a consistent unit of change?

joostjager commented on channeldb/payment_control_test.go:
```diff
@@ -963,3 +973,30 @@ func assertPaymentIndex(t *testing.T, p *PaymentControl,
+
+// assertNoIndex checks that an index for the sequence number provided does not
+// exist.
+func assertNoIndex(t *testing.T, p *PaymentControl, seqNr uint64) {
```
Same comment about using a more black box approach

joostjager commented on channeldb/payments_test.go:
```diff
@@ -422,3 +423,63 @@ func TestQueryPayments(t *testing.T) {
+
+// appendDuplicatePayment adds a duplicate payment to an existing payment. Note
+// that this function requires a unique sequence number.
+func appendDuplicatePayment(t *testing.T, db *DB, paymentHash lntypes.Hash,
```
There is quite a bit of code in this PR to deal with duplicate payments. I am wondering for how long we are going to invest time in those? Maybe there is some kind of middle ground where we don't delete them completely, but only make them retrieval through a special endpoint. They would then also not need the indexing changes and all the associated test code

carlaKC replied:
I think it's worth dealing with legacy in a way that as as seamless to the user as possible, extra endpoints feels like another thing for people to load up into their mental model of how lnd works. If we want to sunset duplicates, I think we'd at least need to communicate that pretty widely, then perhaps have a release which deletes and csv-dumps those payments for the user. We could go down that road, but I think it's a long term plan, and people are having issues with their `ListPyayments` endpoint now, so I'd argue for tolerating them once more and then making a proper deprecation plan.

joostjager commented on channeldb/paginate.go:
```diff
@@ -78,6 +78,14 @@ func (p paginater) cursorStart() ([]byte, []byte) {
+			// index offset we have been provided is beyond our last
+			// item. Since we are paginating backwards, we simply
+			// start at the last item.
+			if indexKey == nil {
```
Not part of test case?

carlaKC replied:
This behaviour is pre-existing for our invoices pagination, so I wanted to separate it out. The previous commit moves the logic across as is, and this one improves it. But I think a better way would be to add the failing test then fix it, going to switch the order around.

carlaKC commented:
>  I think that ideally it should be possible to run every commit on its own without leaving the db in an unusable state for example.

Squashed and re-ordered a bit. Only one that's questionable is the first two commits (add index/ delete index when we delete payments). I can squash them so that everything is done in one, but that makes for a pretty big commit. 

I think the likelihood of us splitting these commits is pretty low (we wouldn't split cherry pick half way through a migration), but it really comes down to reviewer preference so happy to go either way (could even squash when we're closer to merge?)

cfromknecht commented on channeldb/invoice_test.go:
```diff
@@ -877,6 +877,18 @@ func TestQueryInvoices(t *testing.T) {
 			// still pending.
 			expected: pendingInvoices[len(pendingInvoices)-15:],
 		},
+		// Fetch all invoices paginating backwards, with an index offset
```
i too noticed this when working on #4298, good to have this behavior tested 👍

carlaKC commented on channeldb/migration16/migration.go:
```diff
@@ -0,0 +1,153 @@
+	for _, index := range indexList {
+		// Write indexes for each of our sequence numbers.
+		for _, duplicate := range index.sequenceNumbers {
+			err := putIndex(bucket, duplicate, index.paymentHash)
```
Why would we need this in addition to the check in `putIndex`?

carlaKC commented:
> Did you test this manually with duplicates? 

I haven't yet, was thinking about manually inserting a duplicate into a local DB then running the migration, but that probably isn't the same as a real legacy db.

> For commit structure, I'd squash the first three commits into one. 

Squashed 👌

cfromknecht commented on channeldb/payment_control.go:
```diff
@@ -183,6 +214,68 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+		return lntypes.Hash{}, err
+	}
+
+	hash, err := lntypes.MakeHash(paymentHash[:])
```
good point, that is still tbd as to the best way to do that. more than likely we will start to index payments by payment address since that will be a uniform identifier that also works for MPP. when you start getting to recurring payments we would need to index by set_id in order to differentiate logical payments to the same invoice. 

this could be implemented by adding a type byte to the index to signify which index the following 32-byte correspond to, but i don't know how much effort it would be for this PR. perhaps the best approach would be to deal with these details when we get there, and say that `len(value) == 32` means payment hash. when we add the newer indexes, they could receive this type byte and be 33 bytes. i wouldn't want to block progress on this for some undecided details on AMP, but all in all i think we should be able to extend it as we need 👍

carlaKC commented:
@guggero has some duplicates on his node and managed a successful dry run:

```
2020-06-04 12:34:52.208 [WRN] LTND: open /root/.lnd/lnd.conf: no such file or directory
2020-06-04 12:34:52.209 [INF] LTND: Version: 0.10.99-beta commit=v0.5-beta-4510-gb969eb4420354cc13f9d434da93fa093b64529ac, build=production, logging=default
2020-06-04 12:34:52.209 [INF] LTND: Active chain: Bitcoin (network=mainnet)
2020-06-04 12:34:52.209 [INF] LTND: Opening the main database, this might take a few minutes...
2020-06-04 12:34:52.487 [INF] CHDB: Checking for schema update: latest_version=16, db_version=13
2020-06-04 12:34:52.487 [INF] CHDB: Performing database schema migration
2020-06-04 12:34:52.487 [INF] CHDB: Applying migration #14
2020-06-04 12:34:52.487 [INF] CHDB: Creating top-level bucket: "pay-addr-index" ...
2020-06-04 12:34:52.487 [INF] CHDB: Created top-level bucket: "pay-addr-index"
2020-06-04 12:34:52.487 [INF] CHDB: Applying migration #15
2020-06-04 12:34:52.487 [INF] CHDB: Creating top-level bucket: "payments-index-bucket" ...
2020-06-04 12:34:52.487 [INF] CHDB: Created top-level bucket: "payments-index-bucket"
2020-06-04 12:34:52.487 [INF] CHDB: Applying migration #16
2020-06-04 12:34:52.487 [INF] CHDB: Migrating payments to add sequence number index
2020-06-04 12:34:52.494 [INF] LTND: dry run migration successful, exiting
2020-06-04 12:34:52.494 [INF] LTND: Shutdown complete
```

cfromknecht commented on channeldb/payment_control.go:
```diff
@@ -152,6 +162,27 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+		// our payment hash.
+		err = createPaymentIndexEntry(tx, sequenceNum, info.PaymentHash)
+		if err != nil {
+			return err
```
it looks like `errNoIndexBucket` can leak out of the public API even though it is unexported?

cfromknecht commented on channeldb/payment_control.go:
```diff
@@ -183,6 +214,68 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+
+// fetchPaymentIndexEntry gets the payment hash for the sequence number provided
+// from our payment indexes bucket.
+func (p *PaymentControl) fetchPaymentIndexEntry(sequenceNumber uint64) (
```
is this method only used for testing?

cfromknecht commented on channeldb/payment_control.go:
```diff
@@ -74,6 +76,14 @@ var (
+
+	// errNoIndex is returned when an attempt to lookup a payment index is
+	// made for a sequence number that is not indexed.
+	errNoIndex = errors.New("index does not exist")
```
ambiguous as to _which_ index/bucket is missing, e.g. payment index (also above)

cfromknecht commented on channeldb/payment_control.go:
```diff
@@ -183,6 +214,68 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+
+	indexes := tx.ReadWriteBucket(paymentsIndexBucket)
+	if indexes == nil {
+		return errNoIndexBucket
```
can't we assume this index always exists bc of the migration?

cfromknecht commented on channeldb/payment_control.go:
```diff
@@ -183,6 +214,68 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+	if err := kvdb.View(p.db, func(tx walletdb.ReadTx) error {
+		indexBucket := tx.ReadBucket(paymentsIndexBucket)
+		if indexBucket == nil {
+			return errNoIndexBucket
```
same here wrt assuming the index exists. perhaps the error is not needed?

carlaKC commented:
> remove `errNoIndexBucket` because we can assume it exists due to the migration

Indeed, removed this err check, don't need it anymore. 

Also added a type enum for payments so that we're future proofed for AMP. I assume we don't need length since we'll always know what our own db types are, and I haven't bothered with splitting de/serialization out by type since we only have a single type (just read it out and santity check for now, future changes can rework to check type then read accordingly).

cfromknecht commented on channeldb/payment_control.go:
```diff
@@ -183,6 +210,58 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
 }
 
+// paymentIndexType indicates the type of index we have recorded in the payment
+// indexes bucket. This type indicates
```
nit: incomplete sentence

cfromknecht commented on channeldb/payment_control.go:
```diff
@@ -183,6 +210,58 @@ func (p *PaymentControl) InitPayment(paymentHash lntypes.Hash,
+
+// paymentIndexTypeHash is a payment index type which indicates that we have
+// created an index of payment sequence number to payment hash.
+const paymentIndexTypeHash paymentIndexType = iota
```
nice addition 😎

we typically assign persisted enums to constants rather than using `iota`, otherwise reordering the definitions can break the mapping. not a blocker though, happy to change since i'll be extending this area in the near future

cfromknecht commented on channeldb/payments.go:
```diff
@@ -104,6 +104,14 @@ var (
 )
 
+var (
+	errNoSequenceNumber        = errors.New("sequence number not found")
```
in the next commit it looks like it's possible for these errors to be returned via the public interface? if so we should consider exposing them as well and adding godocs
